### PR TITLE
Gradients sharing

### PR DESCRIPTION
**Pull request by raver119**
***WIP; DO NOT MERGE;***

This PR adds option to run ParallelWrapper without averaging, via gradients updates sharing between workers. Following implementations were added:
- direct sharing. Updates are shared "as is", without encoding/decoding. Probably should be default mode for PW
- threshold encoding & topN threshold encoding. only portion of updates are shared. These implementations will be used for distributed training.

### CODE REVIEW

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/Model.java [hunk 1]
```diff
@@ -53,6 +53,13 @@
      */
     void setListeners(IterationListener... listeners);
 
+    /**
+     * This method ADDS additional IterationListener to existing listeners
+     *
+     * @param listener
+     */
+    void addListener(IterationListener listener);
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/Model.java]: Good idea. Minor improvement: we might as well make this ```addListeners(IterationListener...)``` and ```addListeners(Collection<?
 extends IterationListener>)```
That would save users having to chain addListener calls for multiple listeners.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java [hunk 2]
```diff
@@ -318,6 +319,21 @@ public void setLabels(INDArray... labels) {
...
 
     /**
+     * This method allows you to specificy GradientsAccumulator instance to be used with this model
+     *
+     * PLEASE NOTE: Do not use this method unless you understand how to use GradientsAccumulator & updates sharing.
+     * PLEASE NOTE: Do not use this method on standalone model
+     *
+     * @param accumulator
+     */
+    public void setGradientsAccumulator(GradientsAccumulator accumulator) {
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java]: Random thought with these sorts of methods: should we introduce our own annotations (probably defined in ND4J and used everythere) for these sorts of things?
Similar to what Spark does: https://spark.apache.org/docs/1.5.2/api/java/org/apache/spark/annotation/package-summary.html

It'd let people know, and (if easy/possible) it should automatically add warnings to (html/rendered) javadoc etc.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java [hunk 3]
```diff
@@ -363,6 +366,15 @@ public INDArray activate(boolean training) {
         applyDropOutIfNecessary(training);
 
         INDArray z = preOutput(training);
+
+        if (training) {
+            try (MemoryWorkspace wsB = Nd4j.getWorkspaceManager()
+                    .getWorkspaceForCurrentThread(ComputationGraph.workspaceExternal).notifyScopeBorrowed()) {
+                //preOutput = z.unsafeDuplication();
+                //Nd4j.getExecutioner().commit();
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java]: What's the idea here? seems like a no-op currently?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/StochasticGradientDescent.java [hunk 4]
```diff
@@ -37,6 +38,7 @@
  * No line search
  * @author Adam Gibson
  */
+@Slf4j
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/StochasticGradientDescent.java]: Not used?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/BasicGradientsAccumulator.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * This class provides accumulation for gradients for both input (i.e. updates coming from network) and output (comint from one ore more models training at the same time)
+ *
+ * @author raver119@gmail.com
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/BasicGradientsAccumulator.java]: Extra JavaDoc might be good here: it's basic, but in what sense? Compare to what? How does it work?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/BasicGradientsAccumulator.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            barrier.await();
+            if (firstOne.get() == Thread.currentThread().getId()) {
+                // one thread just nullifies this array
+                updates.assign(0.0);
+                hasSomething.set(false);
+
+                firstOne.set(-1L);
+            }
+
+            updatesLock.readLock().unlock();
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/BasicGradientsAccumulator.java]: What happens if anything in this try block throws an exception before .unlock()?
Same thing for other uses of locks/barriers in this class.
It's pretty unlikely, but not impossible.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/MessageHandler.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+     */
+    void initialize(GradientsAccumulator accumulator);
+
+    /**
+     * This method does broadcast of given update message across network
+     *
+     * @param updates
+     * @return TRUE if something was sent, FALSE otherwise
+     */
+    boolean broadcastUpdates(INDArray updates);
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/MessageHandler.java]: If implementations may be using network comm - interface should have error handling? (like IOException etc?)

File: deeplearning4j-scaleout/deeplearning4j-scaleout-parallelwrapper/src/main/java/org/deeplearning4j/parallelism/ParallelWrapper.java [hunk 8]
```diff
@@ -62,14 +86,17 @@
...
+    protected Object[] trainerContextArgs;
+
+    @Getter @Setter protected GradientsAccumulator gradientsAccumulator;
 
     private MagicQueue mq;
 
     // log uncaught exceptions
     Thread.UncaughtExceptionHandler handler = new Thread.UncaughtExceptionHandler() {
         public void uncaughtException(Thread th, Throwable ex) {
             log.error("Uncaught exception: " + ex);
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/deeplearning4j-scaleout-parallelwrapper/src/main/java/org/deeplearning4j/parallelism/ParallelWrapper.java]: It doesn't really matter and I'm being pedantic here, but generally with logging we should use ```log.error("Uncaught exception", ex);```
https://www.slf4j.org/faq.html#paramException

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/Model.java (hunk 1 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/Model.java]: Mmmm, yes, makes sense

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java (hunk 2 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java]: All up for that. API gets more and more complex, so any hints for end-user expectations are more then welcome

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java (hunk 3 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java]: That's preOutput cache draft, i'm playing with it at this moment.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/BasicGradientsAccumulator.java (hunk 6 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/BasicGradientsAccumulator.java]: model dies due to jvm crash.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/MessageHandler.java (hunk 7 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/MessageHandler.java]: Agreed.

File: deeplearning4j-scaleout/deeplearning4j-scaleout-parallelwrapper/src/main/java/org/deeplearning4j/parallelism/ParallelWrapper.java (hunk 8 above)
Comment (raver119) [inline deeplearning4j-scaleout/deeplearning4j-scaleout-parallelwrapper/src/main/java/org/deeplearning4j/parallelism/ParallelWrapper.java]: Not my code, sorry. But i'm going to revisit this thing during spark integration. I don't like UncaughtException at all.

**Comment by raver119** [comment]
So, due to https://github.com/deeplearning4j/deeplearning4j/issues/3385 i'm rewriting encoder/decoder algorithm and adding new kernels, to make it suit CUDA better. Current implementation is ok for CPU though.

**Pull request merged**